### ISSUE TITLE
No minimum version of useragent gem in gemspec allows incompatible versions to break secureheaders

### ISSUE THREAD

**Issue by anglinb** [twitter/secureheaders#336]
**tl;dr** `secureheaders` relies on a _relatively_ recent version of the [`useragent` gem](https://github.com/gshutler/useragent) but the [`secure_headers.gemspec`](https://github.com/twitter/secureheaders/blob/48c69ab33ea376e502a90dd09b42f44878431670/secure_headers.gemspec#L19) file does not place any requirements on the version of the `useragent` gem that may have been explicitly installed in a project's Gemfile or by another dependency that the a project uses. 

**Request:** Please help me pick a minimum [`useragent` gem version](https://github.com/gshutler/useragent/releases) to require in the [`secure_headers.gemspec`](https://github.com/twitter/secureheaders/blob/48c69ab33ea376e502a90dd09b42f44878431670/secure_headers.gemspec#L19).

--- 

# Bug Report

### Expected outcome

 - I expected an error when installing `secureheaders` telling me that my `useragent` gem version was incompatible. 

### Actual outcome

 - I installed `secureheaders` without having to bump by `useragent` gem version leading to an error during bootstrapping of my rail application. 

### Config

 - I am using the default config. 

### Generated headers
 - N/A

### Example Code

```ruby
require 'rails/all'

module MyApp
  class Application < Rails::Application
    SecureHeaders::Configuration.default
  end
end
```
**Gemfile**
```
source "https://rubygems.org"

gem 'rails', '4.0.12'
gem "secure_headers", "~> 3.5.6"
gem 'useragent'
```

**Gemfile.lock** (Excerpt)

```
GEM
  remote: https://rubygems.org/
  specs:
    rails (4.0.12)
      actionmailer (= 4.0.12)
      actionpack (= 4.0.12)
      activerecord (= 4.0.12)
      activesupport (= 4.0.12)
      bundler (>= 1.3.0, < 2.0)
      railties (= 4.0.12)
      sprockets-rails (~> 2.0)
    secure_headers (3.6.5)
      useragent
    useragent (0.6.0)
```

### Stack Trace

```
NameError: uninitialized constant UserAgent::Browsers::Base
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/headers/content_security_policy.rb:27:in `initialize'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/headers/policy_management.rb:195:in `new'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/headers/policy_management.rb:195:in `make_header'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:316:in `block in generate_csp_headers_for_config'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:315:in `each'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:315:in `generate_csp_headers_for_config'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:308:in `generate_csp_headers'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:295:in `cache_headers!'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:74:in `add_configuration'
/usr/local/bundle/gems/secure_headers-3.6.5/lib/secure_headers/configuration.rb:18:in `default'
/myapp/config/application.rb:100:in `<class:Application>'

```
### Fix

Specify the minimum version of the `useragent` gem that `secureheaders` requires.

**Comment by jacobbednarz** [comment]
Fully support this idea, even if it is just `> version` 👍 Are you able to raise a PR to address this?

**Comment by anglinb** [comment]
@jacobbednarz Thanks! 😄  Yes, I will submit a PR--just not sure what version to specify. Might write a script to see which version is the minimum that will work with secureheaders (which version will make the tests pass).

**Comment by anglinb** [comment]
I've confirmed that `0.6.0` **_does not_** work but `0.16.8` **_does_**. Time to binary search? 😜

**Comment by jacobbednarz** [comment]
Oh dear! 

I would probably keep bumping `gem 'useragent'` up from `0.6.0` in your app until the above exception goes away. Or track down where `UserAgent::Browsers::Base` was introduced and go from there.

**Comment by anglinb** [comment]
Yep, I think I'm just going to do it linearly until the tests start passing. Almost reached for my freshman CS class knowledge. Actually, might checkout [this commit](https://github.com/gshutler/useragent/commit/0daf84f68be00d2262864454597255218d92fe07) and just try that. A four year old version (assuming this works) seems like a reasonable requirement.

**Comment by anglinb** [comment]
So here are the results. I tested with both `secureheaders` `v3.5.1` and `v3.5.6` b/c I was using both in different applications. **_All tests_** break when `useragent` is `< v0.9.0`. With `v0.9.0` up to but not including `v0.15.0` break **_one test_**. Therefore, I think we should require `>= v0.15.0`.

**Comment by oreoshake** [comment]
I'm pretty sure this would apply to all 3.x versions.

**Comment by oreoshake** [comment]
I agree a minimum version number is a good thing to have but if this is the only reason it's blowing up, perhaps we could just stop using [`if user_agent.is_a?(UserAgent::Browsers::Base)`](https://github.com/twitter/secureheaders/blob/24340c89b8ceca1060038b36e8278393e90154e0/lib/secure_headers/headers/content_security_policy.rb#L28). 

The purpose of that line appears to be to prevent double parsing or parsing something that can't be parsed. We could instead check whether or not it's a string.

**Comment by oreoshake** [comment]
Oh, the test failures.

**Comment by oreoshake** [comment]
I cherry-picked this into the 3.x line and [released 3.6.6](https://rubygems.org/gems/secure_headers/versions/3.6.6). We haven't cut a 4.x gem yet but this will be included.
